Bluetooth: L2CAP: Remove use of k_sem for credits

Vudentz opened the pull request:
With the changes that introduced a queue k_sem is only used with
K_NO_WAIT which means it is no longer possible to wait/block for credits
so the usage of k_sem is no longer needed and can be safely replaced
with atomic_t just to count the available credits at a given instant.

Fixes #19922

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: L2CAP: Remove use of k_sem for credits

With the changes that introduced a queue k_sem is only used with
K_NO_WAIT which means it is no longer possible to wait/block for credits
so the usage of k_sem is no longer needed and can be safely replaced
with atomic_t just to count the available credits at a given instant.

Fixes #19922

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

jhedberg commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1256,6 +1252,25 @@ static void l2cap_chan_seg_sent(struct bt_conn *conn, void *user_data)
+	k_spin_unlock(&lock, key);
+
+	return true;
+}
```
`l2cap.c` doesn't seem the right place for this. Especially when it uses the reserved `atomic_*` namespace. I'd suggest to split this off into its own patch that adds it to `include/sys/atomic.h`

Vudentz replied:
I was a little hesitant to do that right away since it is not always easy to get things in the kernel side, but perhaps on the atomic.h that would be acceptable.

jhedberg replied:
Btw, the implementation doesn't look correct since it's looking at and decrementing the pointer rather than the value behind it

Vudentz replied:
I experiement a little bit with it residing on atomic.h, the problem is that there is no built-in for it so it needs to be inline and add includes for spinlock, so I don't think it is very clean:

https://gist.github.com/Vudentz/a6b7a1cf93d4e984641107293f6ee900

jhedberg replied:
> I experiement a little bit with it residing on atomic.h, the problem is that there is no built-in for it so it needs to be inline and add includes for spinlock, so I don't think it is very clean

I've added @andrewboie and @andyross as reviewers (marked as codeowners for atomic.h) who can hopefully provide some guidance here. Why do you need to have duplicate implementations in the header and c-file? Why can't you just have a prototype in the header and the implementation in the c-file (which could then use the same lock as other functions there)?

andrewboie replied:
@jhedberg so there's three options for atomic ops, selected in a menu in kernel/Kconfig:

- CONFIG_ATOMIC_OPERATIONS_BUILTIN - This uses the compiler's builtin functions for atomic operations. This is the preferred option, but not all compilers have builtins so we have to fallback on some other options.
- CONFIG_ATOMIC_OPERATIONS_CUSTOM - The arch code has custom assembly routines that Zephyr maintains for these functions.
- CONFIG_ATOMIC_OPERATIONS_C - The worst option. These grab a spinlock and then just do the math non-atomically. Because spinlocks are reserved for kernel use, user mode has to make syscalls. Pretty much for CPUs that lack native atomic instructions.

If you look at the #ifdefery in the header you will see how these implementations are managed, the builtin option has these as simple inlines which translate to the GCC atomics.

andrewboie replied:
BTW https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html GCC does not have a builtin with "test and dec" semantics so I urge you guys to consider re-doing your code to use other atomic ops to get what you need.

jhedberg reviewed (changes requested):
Looks generally fine, except the for wrong placement of the new atomic function

mped-oticon commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1256,6 +1252,25 @@ static void l2cap_chan_seg_sent(struct bt_conn *conn, void *user_data)
+
+	key = k_spin_lock(&lock);
+
+	if (!(*target)) {
```
I realize the spin lock is sufficient, but this currently only works because atomic_t is typedef'd to unsigned int.
The abstraction will not break in future if we use atomic_get, atomic_dec (below).

mped-oticon reviewed (approved):
LGTM. Thanks!

joerchan commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1256,6 +1252,25 @@ static void l2cap_chan_seg_sent(struct bt_conn *conn, void *user_data)
+	static struct k_spinlock lock;
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&lock);
```
could we use atomic_get and atomic_cas and not have the spin-lock?

Vudentz replied:
We could indeed do that, or a get follow by atomic_dec but then it is not an atomic operation but 2 so there is a window where it could be preempted.

joerchan replied:
The window should be handled by having the atomic_cas in a loop until it succeeds AFAIK.

Vudentz replied:
I guess we don't want a busy loop though otherwise it is not that different than using a spin_lock.

andrewboie replied:
You need to delete this function and do something else, working with the existing atomic operations in Zephyr.

joerchan commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1594,7 +1607,7 @@ static void l2cap_chan_update_credits(struct bt_l2cap_le_chan *chan,
 	credits = ((chan->_sdu_len - net_buf_frags_len(buf)) +
 		   (chan->rx.mps - 1)) / chan->rx.mps;
-	credits -= k_sem_count_get(&chan->rx.credits);
+	credits -= atomic_get(&chan->rx.credits);
```
Unrelated to this PR, but credits is s16_t above? Shouldn't it be u16_t @Vudentz ?

Vudentz replied:
I will check about it, I recall it being change to s16_t for some reason but maybe Im mistaken.

mped-oticon replied:
I hope it is u16, since I witness some Androids granting 0xffff credits - and they don't mean -1 credits :)

Vudentz replied:
This code is only used to restore RX credits, so it is not influenced by the remote, it restore the amount based on the SDU and then subtract to the existing credits to know if more credits would be needed to complete the SDU, the result can be negative which is why we are using s16_t, we could use a s32_t though that could overflow so it needs to do a bound check as well.

Vudentz replied:
In the end I just did the bound check separately and used u16_t as it should be.

joerchan reviewed (commented):
So this will replace #21072 then?

tyrantlizard77 reviewed (changes requested):
NACK on implementing non-standard atomic APIs using spinlocks

Vudentz commented:
> NACK on implementing non-standard atomic APIs using spinlocks

Done.

andrewboie commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1256,6 +1252,17 @@ static void l2cap_chan_seg_sent(struct bt_conn *conn, void *user_data)
+	if (!atomic_get(target)) {
+		return false;
+	}
+
```
I think we have a race condition here in between the check to see if target is nonzero, and decrementing the target, you could get preempted in between, or it gets updated on another CPU at the same time.

I recommend you look at lib/os/sem.c, note the use of atomic_cas() with while loops.

Vudentz replied:
Interesting if such construction is recommended why we don't we promote them to atomic_bound_dec and atomic_bound_inc? They could be used as is in this case.

andyross replied:
(Oops, fat-fingered the resolve button, sorry).

To be clear: that construction is required.  Consider the case where you enter this function, decide that the count is non-zero, and then between the if() block and the atomic_dec() another thread runs and decrements the count to zero.  This line will then wrap it around to 0xffffffff incorrectly.   Two atomic operations used sequentially is not atomic.

I don't think having a bounded add/dec operation in the kernel would be a bad idea at all, though it's a little out of scope of the existing atomic.h API which is really there just to expose common/portable hardware instructions for atomic operations.

andrewboie replied:
the bounded_inc/bounded_dec functions have some special semantics related to how it is used with futexes (bounded dec will decrement to minimum -1 and not minimum).

So I'd be very careful in making sure that this works precisely the way you want if you use that code unmodified.

However, looping with atomic_cas() is a general design pattern to make sure that the value wasn't changed underneath you while you were working on it, so whatever you come up with I think would surely look similar.

Vudentz replied:
Yep, Ive added a pretty similar loop, it just doesn't take a minimum value since all we care is that it doesn't decrement bellow zero.

Vudentz commented:
@andyross it been taken care now, let me know if there are any other concerns.

Vudentz commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1256,6 +1252,22 @@ static void l2cap_chan_seg_sent(struct bt_conn *conn, void *user_data)
+
+	return true;
+}
+
```
@andrewboie @andrewboie this would be the equivalent to i.e. atomic_bound_dec(target, 1) which afaik we could inline in atomic.h, but Im fine either way.

Vudentz committed:
Bluetooth: L2CAP: Fix using s16_t to represent credits

Credits are 2 octects long so an s16_t positive portion can only half to
the theorical maximum number of credits, so instead this uses u16_t and
do a bound check instead of checking for negative values.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

tyrantlizard77 reviewed (approved):
LGTM thanks!

andrewboie commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1226,7 +1222,7 @@ static struct net_buf *l2cap_chan_create_seg(struct bt_l2cap_le_chan *ch,
+	if (!atomic_get(&ch->tx.credits) ||
 	    (k_fifo_is_empty(&ch->tx_queue) && !ch->tx_buf)) {
 		return;
 	}
```
one question: any complications here if credits gets set to 0 by another CPU in between the atomic_get() and the k_work_submit() call?

Vudentz replied:
There shouldn't be any problem since the code checks again if there any credits on the work itself, that is actually where test_and_dec is used so it would basically just defered itself again until there are credits available.

andrewboie replied:
sounds good

andyross reviewed (approved):
This looks correct, thanks.

joerchan commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1256,6 +1252,22 @@ static void l2cap_chan_seg_sent(struct bt_conn *conn, void *user_data)
+		}
+
+		new_value = old_value - 1;
+	} while (atomic_cas(target, old_value, new_value) == 0);
```
Minor: I would think while(!atomic_cas()); but I see that atomic_cas has different return values based on ifdefs, and that both versions are used elsewhere in the code.

andrewboie replied:
`while(!atomic_cas())` and `while (atomic_cas() == 0)` are the same. MISRA-C wants us to not treat integers as booleans, so the specific == 0 is added.

joerchan reviewed (approved):
This was a very nice change.

jhedberg commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -797,9 +795,7 @@ static void l2cap_chan_rx_give_credits(struct bt_l2cap_le_chan *chan,
-	while (credits--) {
-		k_sem_give(&chan->rx.credits);
-	}
+	atomic_sub(&chan->rx.credits, credits);
```
Why are you doing `atomic_sub()` here and `atomic_add()` in `l2cap_chan_tx_give_credits()`. Shouldn't they both be `atomic_add()`?

Vudentz replied:
Will fix that, not sure how this worked for me but it obviously wrong.

jhedberg reviewed (changes requested):
See my inline comment about `atomic_sub()` vs `atomic_add()`
